### PR TITLE
Gzip artifacts

### PR DESCRIPTION
**Pull request by Veetaha**
[Here is the test release](https://github.com/Veetaha/rust-analyzer/releases/tag/2020-06-21)

Change in size:  `~ 25 MB -> ~ 8 MB (gzipped)`

The time to gzip during the dist build takes a somewhat considerable amount of time tho.
Having already compiled artifacts this takes in debug mode:
```
~/dev/rust-analyzer (feat/gzip-binaries) $ time cargo xtask dist
    Finished dev [unoptimized] target(s) in 0.06s
     Running `target/debug/xtask dist`
> cargo build --manifest-path ./crates/rust-analyzer/Cargo.toml --bin rust-analyzer --release
    Finished release [optimized] target(s) in 0.05s
> strip ./target/release/rust-analyzer

real    0m34.331s
user    0m34.245s
sys     0m0.078s
```
In release mode this is much faster:
```
~/dev/rust-analyzer (feat/gzip-binaries) $ time cargo run -p xtask --release -- dist
    Finished release [optimized] target(s) in 0.04s
     Running `target/release/xtask dist`
> cargo build --manifest-path ./crates/rust-analyzer/Cargo.toml --bin rust-analyzer --release
    Finished release [optimized] target(s) in 0.06s
> strip ./target/release/rust-analyzer

real    0m2.401s
```

**[UPD]** adding a profile override for `miniz_oxide` does the thing to ensure good performrance

We might need to notify all other ra plugins' maintainers about the change in our GH releases if we merge this PR, or we could leave uncompressed files along with gzipped for a while until everyone migrates.

### CODE REVIEW

File: xtask/src/dist.rs [hunk 1]
```diff
@@ -61,18 +66,28 @@ fn dist_server(nightly: bool) -> Result<()> {
...
     } else if cfg!(target_os = "macos") {
-        ("./target/release/rust-analyzer", "./dist/rust-analyzer-mac")
+        ("macos", "")
     } else {
         panic!("Unsupported OS")
     };
 
-    fs2::copy(src, dst)?;
+    let src = format!("./target/release/rust-analyzer{}", ext);
+    let dest = format!("./dist/rust-analyzer-{}.tar.gz", platform);
```
Comment (bjorn3) [inline xtask/src/dist.rs:80]:
```suggestion
    let dest = format!("./dist/rust-analyzer-{}.gz", platform);
```

It is not a tarball.

Comment (Veetaha) [reply xtask/src/dist.rs:80]: Hmm, I need to recap on this :)

File: editors/code/src/main.ts [hunk 2]
```diff
@@ -291,15 +295,21 @@ async function getServer(config: Config, state: PersistentState): Promise<string
     }
 
     const release = await fetchRelease(config.package.releaseTag);
-    const artifact = release.assets.find(artifact => artifact.name === binaryName);
+    const artifact = release.assets.find(artifact => artifact.name === `rust-analyzer-${platform}.tar.gz`);
```
Comment (bjorn3) [inline editors/code/src/main.ts:298]:
```suggestion
    const artifact = release.assets.find(artifact => artifact.name === `rust-analyzer-${platform}.gz`);
```

**Comment by Veetaha** [comment]
Hmm, actually I think we might need to compile xtask in release mode.
Because when running `carg xtask dist` in release mode it takes 10 times fewer time.

**Comment by lnicola** [comment]
I guess we could use a profile override to compile flate2 with optimizations.

**Comment by Veetaha** [comment]
Hmm, that's a nice idea

**Comment by Veetaha** [comment]
Squashed the changes

**Comment by killercup** [comment]
FTR, I was about to say that github most likely already uses deflate as HTTP encoding -- but then I checked and according to `curl --compressed  -L https://github.com/rust-analyzer/rust-analyzer/releases/download/nightly/rust-analyzer-mac -w '%{size_download}' -so /dev/null` it doesn't.

**Comment by matklad** [comment]
Not sure if we need to do this -- I'd rather keep this minimal. The end-goal is distribution via rustup, so additional complexity here won't be worthwhile over the long run. This will also make life harder for folks who maintain other editor plugins.

**Comment by matklad** [comment]
What would help long term is pushing this PR over the finish line: https://github.com/rust-lang/rust/pull/72978 (it currently misses the bit about vendoring).

**Comment by lnicola** [comment]
@matklad do you plan to keep offering binary releases after https://github.com/rust-lang/rust/pull/72978?

**Comment by matklad** [comment]
@lnicola for at least medium term -- yes

**Comment by matklad** [comment]
Hm, I am thinking that maybe we should enable `debug = 1` for release profile? Lack of line numbers from reports is frustraiting. I guess if we do that, it might make sense to enable compression

**Comment by lnicola** [comment]
Do we even get backtraces? I think we strip the binaries.

**Comment by lnicola** [comment]
Maybe we should enable debug info for the binary releases, but not for source builds.

**Comment by Veetaha** [comment]
@matklad this is the reason we don't strip binaries for nightlies, haven't found a PR, but here is the commit: https://github.com/rust-analyzer/rust-analyzer/commit/ffb7ea678b66c7407671361b1ee77bd9d2d6616c

**Comment by matklad** [comment]
So that nightly users have sane stack traces. I think it makes sense to
extend this to everybody, if, after gziping, the binary size is within
reason.

**Comment by Veetaha** [comment]
Removed stripping.
Also, I am leaving bare binaries output from `cargo xtask dist`  because removing non-compressed binaries from Github releases is a breaking change for other plugins that download our releases (coc.nvim rust-analyzer).
You can see the updated example release in my fork repo: https://github.com/Veetaha/rust-analyzer/releases/tag/2020-07-01
This is strange that linux binaries are 40MB unzipped, but windows/mac ones are only 32/33MB...
cc @matklad @lnicola

File: xtask/src/dist.rs [hunk 3]
```diff
@@ -46,17 +52,14 @@ fn dist_client(version: &str, release_tag: &str) -> Result<()> {
...
-        std::env::set_var("CC", "clang");
+        env::set_var("CC", "clang");
         run!(
             "cargo build --manifest-path ./crates/rust-analyzer/Cargo.toml --bin rust-analyzer --release"
             // We'd want to add, but that requires setting the right linker somehow
             // --features=jemalloc
         )?;
-        if !nightly {
-            run!("strip ./target/release/rust-analyzer")?;
-        }
```
Comment (bjorn3) [inline xtask/src/dist.rs:59]: :tada:

**Comment by Veetaha** [comment]
I also wonder whether we really need to have a special case `.exe` extension on windows?

**Comment by Veetaha** [comment]
Tested this on host Ubuntu and VirtualBox windows, both work.

**Comment by bjorn3** [comment]
```bash
$ size -A -x rust-analyzer-linux
section                   size        addr
.interp                   0x1c       0x270
.note.ABI-tag             0x20       0x28c
.note.gnu.build-id        0x24       0x2ac
.gnu.hash               0x523c       0x2d0
.dynsym                0x13c68      0x5510
.dynstr                0x4b686     0x19178
.gnu.version            0x1a5e     0x647fe
.gnu.version_r           0x190     0x66260
.rela.dyn              0xe1900     0x663f0
.init                     0x1a    0x147cf0
.plt                      0x10    0x147d10
.plt.got                  0xf0    0x147d20
.text                0x16dc110    0x147e10
.fini                      0x9   0x1823f20
.rodata                0xe5de3   0x1823f40
.eh_frame_hdr          0x6f574   0x1909d24
.eh_frame             0x33ca64   0x1979298
.gcc_except_table     0x17f8dc   0x1cb5cfc
.tdata                    0x60   0x2035700
.tbss                    0x458   0x2035760
.init_array               0x10   0x2035760
.fini_array                0x8   0x2035770
.jcr                       0x8   0x2035778
.data.rel.ro           0x98c10   0x2035780
.dynamic                 0x220   0x20ce390
.got                    0x6a38   0x20ce5b0
.data                    0x209   0x20d5000
.bss                     0xf08   0x20d5210
.comment                  0x7f         0x0
.debug_aranges          0x8860         0x0
.debug_pubnames        0x31761         0x0
.debug_info            0xa1f13         0x0
.debug_abbrev           0x197e         0x0
.debug_line            0x5ee2f         0x0
.debug_frame             0xf18         0x0
.debug_str             0xa9bdf         0x0
.debug_loc             0x27d48         0x0
.debug_macinfo            0x19         0x0
.debug_pubtypes          0x132         0x0
.debug_ranges          0x6ea70         0x0
Total                0x2153c8d

$ size -A -x ~/Downloads/rust-analyzer-linux | grep debug | tr -s ' ' | cut -d' ' -f2 | tr '\n' '+'
0x8860+0x31761+0xa1f13+0x197e+0x5ee2f+0xf18+0xa9bdf+0x27d48+0x19+0x132+0x6ea70+
$ echo $((0x8860+0x31761+0xa1f13+0x197e+0x5ee2f+0xf18+0xa9bdf+0x27d48+0x19+0x132+0x6ea70)) # debuginfo (sum(.debug_*))
2611835
# = 0x27da7b = ~2.5MiB
$ echo $((0x6f574+0x33ca64+0x17f8dc)) # unwind info (.eh_frame_hdr + .eh_frame + .gcc_except_table)
5421236
# 0 0x52b8b4 = ~5.2MiB
$ echo $((0x2153c8d-0x16dc110-0xe5de3)) # total - .text - .rodata
10034586
# = 0x991d9a = ~9.6MiB
```

Most of the non code/data part seems to be debuginfo and unwind info.

**Comment by Veetaha** [comment]
@bjorn3 is it possible to get these stats for windows/mac?

**Comment by bjorn3** [comment]
For MachO files `objdump --headers` should work I think. It has a `Size` column for every section. If you add `| grep "\."` then flags are filtered out, reducing visual clutter of the output.

**Comment by lnicola** [comment]
```
$ llvm-objdump --headers rust-analyzer-windows.exe                                                                                                                                                   

rust-analyzer-windows.exe:	file format COFF-x86-64

Sections:
Idx Name          Size     VMA              Type
  0 .text         019b581c 0000000140001000 TEXT
  1 .rdata        007129f6 00000001419b7000 DATA
  2 .data         00000c00 00000001420ca000 DATA
  3 .pdata        0019a0dc 00000001420cc000 DATA
  4 .reloc        00011b68 0000000142267000 DATA
```

**Comment by matklad** [comment]
bors d+

(some conflicts in cargo.lock)

(miniz_oxide will be deduped after https://github.com/rust-lang/backtrace-rs/pull/354)

**Comment by matklad** [comment]
#5250 updates miniz

**Pull request merged**